### PR TITLE
Add C++ module for `xmake project`

### PR DESCRIPTION
**Pull request by glcraft**
Add C++ modules (from `c++.build.modules` rule) as built sourcebatch 
Add required flags to modules to get the correct command line generated with `xmake project`

### CODE REVIEW

**Comment by glcraft** [comment]
## Note
Right now, `xmake project -k compile_commands` works good. But `xmake build` won't works.
Since I try to add build commands in `c++.build.modules` (in addition to `c++.build.modules.builder`) something is wrong. 

When i run `xmake b`, xmake tries to build something with no input file nor output file...

I noticed that in `before_buildcmd_files` of the rule `c++.build.modules`, it happens because of this line : 
```lua
common.build_modules_for_batchcmds(target_clone, batchcmds, sourcebatch, modules, opt)
```
If I remove this line, `xmake b` and `xmake project` works. But there is the problem that modules are not build before `xmake project` call, so the required flags in the compile command line are not specified, leading in errors by the LSP.

File: xmake/rules/c++/modules/xmake.lua [hunk 1]
```diff
@@ -57,6 +57,54 @@ rule("c++.build.modules")
             target:data_set("cxx.has_modules", true)
         end
     end)
+    before_buildcmd_files(function(target, batchcmds, sourcebatch, opt)
```
Comment (waruqi) [inline xmake/rules/c++/modules/xmake.lua:60]: It also affects module builds, so please don't add it to c++.build.modules!

File: xmake/plugins/project/clang/compile_commands.lua [hunk 2]
```diff
@@ -38,7 +38,7 @@ end
 function _sourcebatch_is_built(sourcebatch)
     -- we can only use rulename to filter them because sourcekind may be bound to multiple rules
     local rulename = sourcebatch.rulename
-    if rulename == "c.build" or rulename == "c++.build"
+    if rulename == "c.build" or rulename == "c++.build" or rulename == "c++.build.modules"
```
Comment (waruqi) [inline xmake/plugins/project/clang/compile_commands.lua:41]: you can use `"c++.build.modules.builder"`

**Comment by glcraft** [comment]
By using `"c++.build.modules.builder"`, it is better. `xmake b` and `xmake project -k compile_command` works with clang now.
I applied the change for msvc but idi not tried yet

**Comment by waruqi** [comment]
Is it ready now?

**Comment by glcraft** [comment]
No it's not
I tried on Windows with msvc, and it seems to work fine with "c++.build" only : modules are correctly specified. By adding "c++.build.modules.builder" in compiled source batch in the plugin, there are doubles with missing flags in the compile_commands 
I don't know how to solve this...

**Comment by waruqi** [comment]
it will be added here.

https://github.com/xmake-io/xmake/blob/e45896ac7e84dbaa88c708584f389d6eacc1dd53/xmake/plugins/project/clang/compile_commands.lua#L240
https://github.com/xmake-io/xmake/blob/e45896ac7e84dbaa88c708584f389d6eacc1dd53/xmake/plugins/project/clang/compile_commands.lua#L156

**Comment by waruqi** [comment]
Try it again.

**Comment by waruqi** [comment]
It works for me. gcc/clang/msvc

### gcc

```
[
{
  "directory": "/mnt/xmake/tests/projects/c++/modules/hello",
  "arguments": ["/usr/bin/gcc", "-m64", "-fvisibility=hidden", "-fvisibility-inlines-hidden", "-O3", "-std=c++20", "-fmodules-ts", "-fmodule-mapper=build/hello/mapper.txt", "-D_GLIBCXX_USE_CXX11_ABI=0", "-DNDEBUG", "-x", "c++", "-c", "src/hello.mpp", "-o", "build/.objs/hello/linux/x86_64/release/src/hello.mpp.o"],
  "file": "src/hello.mpp"
},
{
  "directory": "/mnt/xmake/tests/projects/c++/modules/hello",
  "arguments": ["/usr/bin/gcc", "-c", "-m64", "-fvisibility=hidden", "-fvisibility-inlines-hidden", "-O3", "-std=c++20", "-fmodules-ts", "-fmodule-mapper=build/hello/mapper.txt", "-D_GLIBCXX_USE_CXX11_ABI=0", "-DNDEBUG", "-o", "build/.objs/hello/linux/x86_64/release/src/main.cpp.o", "src/main.cpp"],
  "file": "src/main.cpp"
}]
```

### clang

```
[
{
  "directory": "/Users/ruki/projects/personal/xmake/tests/projects/c++/modules/hello",
  "arguments": ["/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain
/usr/bin/clang", "-Qunused-arguments", "-target", "x86_64-apple-macos12.6", "-isysroot", "/Appl
ications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk",
 "-fvisibility=hidden", "-fvisibility-inlines-hidden", "-O3", "-std=c++20", "-fmodules-ts", "-D
NDEBUG", "-fmodules-cache-path=build/.gens/hello/macosx/x86_64/release/rules/modules/cache", "-
x", "c++-module", "--precompile", "-c", "src/hello.mpp", "-o", "build/.gens/hello/macosx/x86_64
/release/rules/modules/cache/2d9b7c0b/hello.pcm"],
  "file": "src/hello.mpp"
},
{
  "directory": "/Users/ruki/projects/personal/xmake/tests/projects/c++/modules/hello",
  "arguments": ["/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain
/usr/bin/clang", "-c", "-Qunused-arguments", "-target", "x86_64-apple-macos12.6", "-isysroot",
"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0
.sdk", "-fvisibility=hidden", "-fvisibility-inlines-hidden", "-O3", "-std=c++20", "-fmodules-ts
", "-DNDEBUG", "-fmodule-file=build/.gens/hello/macosx/x86_64/release/rules/modules/cache/2d9b7
c0b/hello.pcm", "-o", "build/.objs/hello/macosx/x86_64/release/src/main.cpp.o", "src/main.cpp"]
,
  "file": "src/main.cpp"
}]
```

### msvc

```
[
{
  "directory": "C:\\Users\\wangrunqing\\AppData\\Local\\.xmake\\service\\server\\remote_build\\
sessions\\c05e40aa\\source",
  "arguments": ["C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14
.37.32705\\bin\\HostX64\\x64\\cl.exe", "/nologo", "/O2", "/fp:fast", "/std:c++20", "/EHsc", "/D
NDEBUG", "/TP", "/c", "src\\hello.mpp", "/Fobuild\\.objs\\hello\\windows\\x64\\release\\src\\he
llo.mpp.obj", "/ifcOutput", "build\\.gens\\hello\\windows\\x64\\release\\rules\\modules\\cache\
\2d9b7c0b\\hello.ifc", "/interface"],
  "file": "src\\hello.mpp"
},
{
  "directory": "C:\\Users\\wangrunqing\\AppData\\Local\\.xmake\\service\\server\\remote_build\\
sessions\\c05e40aa\\source",
  "arguments": ["C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14
.37.32705\\bin\\HostX64\\x64\\cl.exe", "/c", "/nologo", "/O2", "/fp:fast", "/std:c++20", "/EHsc
", "/DNDEBUG", "/TP", "/c", "/reference", "hello=build\\.gens\\hello\\windows\\x64\\release\\ru
les\\modules\\cache\\2d9b7c0b\\hello.ifc", "/Fobuild\\.objs\\hello\\windows\\x64\\release\\src\
\main.cpp.obj", "src\\main.cpp"],
  "file": "src\\main.cpp"
}]
```

**Comment by glcraft** [comment]
I just tried on my clang and msvc, the generator works fine 👍 
I noticed that modules are not compiled after `xmake project`, we have to compile them using `xmake b` to get them available for the LSP.
But I think it's better like that. `xmake project` should not build the module, just add the module file where it should be 👍

**Comment by Arthapz** [comment]
I'll try this after work but seems fine to me

**Pull request merged**